certstore_linux.go

```diff
@@ -1,14 +1,324 @@
+	C.memcpy(unsafe.Pointer(digestC.data), unsafe.Pointer(&digest[0]), C.size_t(len(digest)))
+
+	if C.SGN_Digest(key, goToNssAlg[hash], resultC, digestC) != C.SECSuccess {
+		if C.PORT_GetError() == C.SEC_ERROR_INVALID_ALGORITHM {
+			return nil, ErrUnsupportedHash
+		}
+		return nil, fmt.Errorf("error signing: %s\n", C.GoString(C.GetErrorString()))
+	}
+
+	if key.keyType == C.ecKey {
+		var dsaSig C.SECItem
+		if C.DSAU_EncodeDerSigWithLen(&dsaSig, resultC, resultC.len) != C.SECSuccess {
+			return nil, errors.New("unable to generate dsa signature")
+		}
+		resultC = &dsaSig
+	}
+	return C.GoBytes(unsafe.Pointer(resultC.data), C.int(resultC.len)), nil
+}
+
+// Import implements the Store interface.
+func (nssStore) Import(data []byte, password string) error {
+	unicodePassword, err := bmpString(password)
+	if unicodePassword == nil {
+		return err
+	}
+	pass := C.SECITEM_AllocItem(nil, nil, C.uint(len(unicodePassword)))
+	if pass == nil {
+		return errors.New("error allocating memory for PKCS#12 password")
+	}
+	defer C.SECITEM_FreeItem(pass, 1)
+	C.memcpy(unsafe.Pointer(pass.data), unsafe.Pointer(&unicodePassword[0]), C.size_t(len(unicodePassword)))
+	p12 := C.SEC_PKCS12DecoderStart(pass, nil, nil, nil, nil, nil, nil, nil)
+	if p12 == nil {
+		return fmt.Errorf("error initialising PKCS#12 decoding: %s", C.GoString(C.GetErrorString()))
+	}
+	decoded := C.SEC_PKCS12DecoderUpdate(p12, (*C.uchar)(unsafe.Pointer(&data[0])), C.size_t(len(data)))
+	if decoded != C.SECSuccess {
+		return fmt.Errorf("error during PKCS#12 decoding: %s", C.GoString(C.GetErrorString()))
+	}
+	authenticated := C.SEC_PKCS12DecoderVerify(p12)
+	if authenticated != C.SECSuccess {
+		return fmt.Errorf("error during PKCS#12 verification: %s", C.GoString(C.GetErrorString()))
+	}
+	validated := C.SEC_PKCS12DecoderValidateBags(p12, (*[0]byte)(C.P12U_NicknameCollisionCallback))
+	if validated != C.SECSuccess {
+		return fmt.Errorf("error during PKCS#12 bag validation: %s", C.GoString(C.GetErrorString()))
+	}
+	imported := C.SEC_PKCS12DecoderImportBags(p12)
+	if imported != 0 {
+		return fmt.Errorf("error during PKCS#12 import: %s", C.GoString(C.GetErrorString()))
+	}
+	return nil
+}
+
+// Close implements the Store interface.
+func (ctx nssStore) Close() {
+	nssCtx := C.NSSInitContext(ctx)
+	C.NSS_ShutdownContext(&nssCtx)
+}
+
+// openStore opens the current user's NSS database.
 func openStore() (Store, error) {
-	return nil, errors.New("certstore only works on macOS and Windows")
+	homeDir, err := os.UserHomeDir()
+
+	if err != nil {
+		return nil, err
+	}
+
+	nssdb := path.Join(homeDir, ".pki", "nssdb")
+	if _, err := os.Stat(nssdb); os.IsNotExist(err) {
+		return nil, fmt.Errorf("NSS database not found at %s", nssdb)
+	}
+
+	var (
+		nssdbC       = C.CString(nssdb)
+		emptyStringC = C.CString("")
+	)
+	defer C.free(unsafe.Pointer(nssdbC))
+	defer C.free(unsafe.Pointer(emptyStringC))
+
+	ctx := C.NSS_InitContext(nssdbC, emptyStringC, emptyStringC, emptyStringC, nil, C.uint(0))
```

Comment:
Another annoying Cgo difference: `C.PRUint32(0)` instead of `C.uint(0)` gets this building on older versions of Go.